spacegraphcats/catlas.py

```diff
@@ -54,42 +57,65 @@ def load(self, level):
-            graph = read_from_gxt(f, radius=1, directed=False, sequential=False)
+            graph = read_from_gxt(f, radius=1, directed=False,
+                                  sequential=False)
             # move back to the beginning of the temporary file and read the
             # catlas
             tmpf.seek(0)
             root = CAtlas.read(tmpf)
             tmpf.close()
-            print("Root has children {}".format([i.vertex for i in root.children]))
+            print("Root has children {}".format(
+                    [i.vertex for i in root.children]))
             nodes = {node.vertex: node for node in root.children}
             idx = root.idx
             level = root.level
 
+            # sanity check that the catlas nodes and graph vertices correspond
             if set(graph.nodes) ^ set(nodes.keys()):
-                print(graph.nodes)
-                print(list(nodes.keys()))
-                raise ValueError("graph should have the same nodes as the previous level")
+                # they are not equal when there are isolated vertices, which
+                # cannot be represented in the edge list file format.  We need
+                # to make sure that these vertices are indeed isolated by
+                # checking that they are not dominated by multiple vertices.
+                #print(graph.nodes)
+                #print(list(nodes.keys()))
+                parent_count = defaultdict(int)
+                for _, node in nodes.items():
+                    for u in node.children:
+                        parent_count[u.vertex] += 1
+                for v in nodes:
+                    if v not in graph and parent_count[v] != 1:
+                        print("{} has {} parents".format(v, parent_count[v]))
+                        raise ValueError("graph should have the same nodes as"
+                                         " the previous level")
+                    else:
+                        graph.add_node(v)
             return graph, nodes, idx, level
 
-    def save(self, graph, nodes, idx, level):
-        """Write out a partial computation."""
-        if self.ignore:
-            return
+    def _save(self, graph, nodes, idx, level):
+        """Method used by the thread to write out."""
         outfile = self.name(level)
         print("Writing to file {}".format(outfile))
-        with open(outfile, 'w') as f:
+        with gzip.open(outfile, 'wt') as f:
             # make a dummy root to write the catlas using catlas.write method
             root = CAtlas(idx, 0, level+1, nodes.values())
             root.write(f)
             f.write("###\n")
             write_to_gxt(f, graph)
+
+    def save(self, graph, nodes, idx, level):
+        """Write out a partial computation."""
+        if self.ignore:
+            return
+        # run on a separate thread because the files can get large
+        thread = Thread(target=self._save, args=(graph, nodes, idx, level))
```

Comment:
Are we sure that we don't change the graph or nodes? If we do, we need to make a deep copy or should not use threading.